### PR TITLE
Fix(overlap_sync_param): implement correct overlap_sync_param

### PR DESCRIPTION
**Pull request by li126com**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

企业微信截图_17024651723915

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: internlm/utils/megatron_timers.py [hunk 1]
```diff
@@ -23,14 +25,16 @@ def start(self, reset_all=True):
...
+            self.stream.synchronize()
         self.start_time = time.time()
         self.started_ = True
 
     def stop(self):
         """Stop the timer."""
         assert self.started_, "timer is not started"
-        self.stream.synchronize()
+        if not gpc.config.hybrid_zero_optimizer.overlap_sync_param:
+            self.stream.synchronize()
```
Comment (mwiacx) [inline internlm/utils/megatron_timers.py:37]: 这个实际测试了，应该不影响，他只同步计算stream，与通信无关。

File: internlm/utils/gputest.py [hunk 2]
```diff
@@ -36,7 +36,8 @@ def empty_cache_and_diag(batch_count, interval=50):
     if interval <= 0:
         interval = 50
 
-    cuda_memory_analyze(batch_count, batch_count % int(interval) == 0 or batch_count <= 5)
+    if not gpc.config.hybrid_zero_optimizer.overlap_sync_param:
+        cuda_memory_analyze(batch_count, batch_count % int(interval) == 0 or batch_count <= 5)
```
Comment (mwiacx) [inline internlm/utils/gputest.py:40]: @SolenoidWGT 国腾看看这个，`cuda_memory_analyze`里面的cuda.sync主要起到什么作用，是在开启overlap_sync_param不进行cuda_memory_analyze比较好，还是解决cuda_memory_analyze里面的cuda sync比较好，个人比较倾向于后者

File: internlm/solver/optimizer/hybrid_zero_optim.py [hunk 3]
```diff
@@ -837,8 +837,7 @@ def _step(self, closure=None, norms=None):
                     fp16_param.data.copy_(fp32_param)
 
         torch.cuda.synchronize()
-        with torch.cuda.stream(self._comm_bcast_stream):
```
Comment (mwiacx) [inline internlm/solver/optimizer/hybrid_zero_optim.py:840]: `self._comm_bcast_stream`属性和实例化，可以清理掉了

File: internlm/utils/gputest.py (hunk 2 above)
Comment (SolenoidWGT) [reply internlm/utils/gputest.py:40]: 我感觉也可以去掉cuda sync，去掉后不影响显存profile的结果，因为这个函数调用时不太会有还在进行中的cuda kernel了

File: internlm/solver/optimizer/utils.py [hunk 4]
```diff
@@ -857,16 +877,35 @@ def __init__(self, model: Union[nn.Module, nn.ModuleList]) -> None:
...
+            # launch broadcast for current layer
+            if current_module == list(self._block_to_param.keys())[0]:
+                self._launch_handle(current_module)
+
+            # if this is not the last layer
+            # launch broadcast for next layer
+            if next_layer:
+                self._launch_handle(next_layer)
+
+            for handle in self._block_to_handles[current_module]:
```
Comment (SolenoidWGT) [inline internlm/solver/optimizer/utils.py:908]: ckpt存储的step时需要禁掉overlap

File: internlm/solver/optimizer/utils.py [hunk 5]
```diff
@@ -817,17 +819,35 @@ def __init__(self, model: Union[nn.Module, nn.ModuleList]) -> None:
...
+                            print(f"{block.__class__.__name__}_layer_{layer_idx}", flush=True)
+                        last_module = block
                 else:
                     # record the block that a parameter belongs to
-                    # self._block_to_param[name] = list(children.parameters())
                     self._block_to_param[children] = list(children.parameters())
+                    if module_idx == 0:
+                        assert "embedding" in f"{children.__class__.__name__}"
+                        assert last_module is None
+                        self._block_next_block[children] = children
```
Comment (SolenoidWGT) [inline internlm/solver/optimizer/utils.py:846]: 846行我加的冗余了，可以去掉

File: internlm/solver/optimizer/utils.py [hunk 6]
```diff
@@ -817,17 +819,35 @@ def __init__(self, model: Union[nn.Module, nn.ModuleList]) -> None:
             if isinstance(_chunk, NaiveAMPModel):
                 _chunk = _chunk.model
 
-            for _, children in _chunk.named_children():
+            # In order to ensure orderliness, non-PackedFlashBaseLayer1D modules are required to
+            # communicate synchronously before the first fwd.
+            # BUG: The order of traversal is not necessarily the order of actual fwd/bwd execution!!!
+            last_module = None
+            for module_idx, children in _chunk.named_children():
```
Comment (SolenoidWGT) [inline internlm/solver/optimizer/utils.py:826]: 需要加一个check，如果model实际fwd执行顺序和遍历nn.module的顺序不一致，需要failed掉，不过目前我们实现的modeling都满足这个约束，这个可以最后搞

**Pull request closed without merging**